### ISSUE TITLE
TypeScript: improve UseCase#execute typing

### ISSUE THREAD

**Issue by azu** [almin/almin#107]
From https://github.com/almin/almin/pull/101#discussion_r103071022

Currently, Following pattern is pass.
We want to improve this.

```ts
class UseCase extends UseCase {
    execute(value: string) {
        // do some
    }
}
const useCase = new UseCase();
// FIXME: This execute(number) is pass!
context.useCase(useCase).execute(42).then(() => {
  
})
```

**Comment by azu** [comment]
Wait for https://github.com/Microsoft/TypeScript/issues/5453

**Comment by azu** [comment]
Can we make it?

```ts
type UseCases = AUseCase | BUseCase;
const context = new Context<UseCases>();
context.useCase(new AUseCase).execute(1,2,3);
```

**Comment by azu** [comment]
## Status 2017-03-29

I have a step-by-step plan to resolve this issue.

**Step 1**. Add generics to `context.useCase<UseCaseArgs>(useCase).execute(arg)`

The user can write following:

```ts
type ParentUseCaseArgs = string;
class ParentUseCase extends UseCase {
    execute(value: ParentUseCaseArgs) {
        return this.context.useCase(new ChildUseCase()).execute(value);
    }
}
context.useCase<ParentUseCaseArgs>(parentUseCase).execute("arg").then(() => {
    const state = context.getState<StoreState>();
    console.log(state.A.a);
    console.log(state.B.b);
}).catch((error: Error) => {
    console.error(error);
});
```

But, it is lengthy syntax.

**Step 2**: if TypeScript implement [Variadic Kinds](https://github.com/Microsoft/TypeScript/issues/5453 "Variadic Kinds"), We can move to that.

The user can write following:

```ts
class ParentUseCase extends UseCase {
    execute(value: string) {
        return this.context.useCase(new ChildUseCase()).execute(value);
    }
}
context.useCase(parentUseCase).execute("arg").then(() => {
    const state = context.getState<StoreState>();
    console.log(state.A.a);
    console.log(state.B.b);
}).catch((error: Error) => {
    console.error(error);
});
```

**Comment by kimamula** [comment]
A current workaround for the variadic kinds I think of is something like the following.

```ts
interface UseCaseBase {
    // define UseCase's methods and properties other than execute
    foo(): string;
}
interface UseCase0 extends UseCaseBase {
    execute(): any;
}
interface UseCase1<T1> extends UseCaseBase {
    execute(arg1: T1): any;
}
interface UseCase2<T1, T2> extends UseCaseBase {
    execute(arg1: T1, arg2: T2): any;
}
// ... create UseCaseX as many as you want

interface UseCaseConstructor {
    new (): UseCase0;
    new<T1> (): UseCase1<T1>;
    new<T1, T2> (): UseCase2<T1, T2>;
}

// Implementation of UseCase
abstract class _UseCase {
    abstract execute(...args: any[]): any;
    foo() {
        return '';
    }
}

// Cast UseCase
const UseCase = _UseCase as UseCaseConstructor;

class SomeUseCase extends UseCase {
    execute() {}
}

class AnotherUseCase extends UseCase<string> {
    execute(arg1: string) {}
}

class YetAnotherUseCase extends UseCase<string, number> {
    execute(arg1: string, arg2: number) {}
}

class Context {
    useCase(useCase: UseCase0): { execute(): Promise<void> };
    useCase<T1>(useCase: UseCase1<T1>): { execute(arg1: T1): Promise<void> };
    useCase<T1, T2>(useCase: UseCase2<T1, T2>): { execute(arg1: T1, arg2: T2): Promise<void> };
    useCase(useCase: { execute(...args: any[]): any }): { execute(...args: any[]): Promise<void> } {
        // todo implement
    }
}
```

http://bit.ly/2opX4l2

What do you think of it?

**Comment by azu** [comment]
@kimamula Thanks for interesting idea.

**Comment by azu** [comment]
Land Variadic types on Roadmap(future).

- https://github.com/Microsoft/TypeScript/wiki/Roadmap#future

**Comment by azu** [comment]
Currently, almin support `executor` method https://github.com/almin/almin/issues/193 
This `executor` is type-safe method.

```js
executor(executor: (useCase: Pick<T, "execute">) => any): Promise<void>;
```

Example:

```ts
context.useCase(new MyUseCase())
 .executor(useCase => useCase.execute("value"))
 .then(() => {
   console.log("test");
 });
```

📝 Document: https://almin.js.org/docs/en/usecaseexecutor-api.html#executorexecutor-usecase-pick-t-execute-any-promise-void

However, `executor` is redundancy.
We will continue to investigate this issue.

**Comment by azu** [comment]
Maybe, We can realive `execute()` typing by conditional types.

```ts
// rough note
export interface UseCase<T>{
  execute(args: T);
}

type ResolveUseCaseArgType<T> =
  T extends UseCase<infer R> ? R : never;

class Context {
    useCase<T extends UseCase>(useCase: T): UseCaseExecutor<T>;
      type UseCaseArgs = ResolveUseCaseArgType<T>;
      return new UseCaseExecutor<UseCaseArgs>(useCase);
}
```


- [Conditional types by ahejlsberg · Pull Request #21316 · Microsoft/TypeScript](https://github.com/Microsoft/TypeScript/pull/21316)
- [Type inference in conditional types by ahejlsberg · Pull Request #21496 · Microsoft/TypeScript](https://github.com/Microsoft/TypeScript/pull/21496)
- [TypeScript 2.8 の Conditional Types について - Qiita](https://qiita.com/Quramy/items/b45711789605ef9f96de)

**Comment by azu** [comment]
https://github.com/almin/almin/issues/107#issuecomment-365166570 can not be realized.

**Comment by azu** [comment]
> feat(almin): Support `context.useCase#execute` typing #342

I've implemented type-check of `execute` method in #342 
But, it is has a drawback again to `executor` method.

It it a limitation of `execute` typing.

### `executor()`

```ts
import { UseCase, Context } from "almin";
class MyUseCaseA extends UseCase {
    execute(_a: string) {}
}
const context = new Context({
    store: createStore({ name: "test" })
});

// valid
context.useCase(new MyUseCaseA()).executor(useCase => useCase.execute("A")); 
// invalid
context.useCase(new MyUseCaseA()).executor(useCase => useCase.execute()); // no argument
context.useCase(new MyUseCaseA()).executor(useCase => useCase.execute(1)); // can not pass number
context.useCase(new MyUseCaseA()).executor(useCase => useCase.execute("A",  42)); // require 1 argument, but specify 2 arguments
```


### `execute()`

```ts
import { UseCase, Context } from "almin";
class MyUseCaseA extends UseCase {
    execute(_a: string) {}
}
const context = new Context({
    store: createStore({ name: "test" })
});

// valid
context.useCase(new MyUseCaseA()).execute("A");
// invalid
context.useCase(new MyUseCaseA()).execute(); // no argument
context.useCase(new MyUseCaseA()).execute(1); // can not pass number
// valid in almin 0.17
// It should be error 
context.useCase(new MyUseCaseA()).execute("A". 42);
```

**Comment by azu** [comment]
TypeScript 3.0 will work completely for `execute` method.

> Extracting and spreading parameter lists with tuples
> https://blogs.msdn.microsoft.com/typescript/2018/07/12/announcing-typescript-3-0-rc/#tuples-and-parameters
